### ISSUE TITLE
Propose Adding a GetOrAdd(TKey key, Func<TKey, TValie> valueFactory) to Dictionary<TKey, TValue>

### ISSUE THREAD

**Issue by SamuelEnglard** [dotnet/corefx#2877]
On `ConcurrentDictionary<TKey, TValue>` there is the very useful function `TValue GetOrAdd(TKey key, Func<TKey, TValie> valueFactory)`. The pattern that the method represents is one that issued commonly on the "normal" `Dictionary<TKey, TValue>` as well. I propose adding such a method for this common use case. Besides simplifying it will also make switching between the two dictionaries simpler.

**Comment by hackcraft** [comment]
Suggested at #1942 within the comment thread too. Possible implementation at https://github.com/hackcraft/coreclr/commit/1a982053ae39ee0e418d370fdee3497ecefd0e5a

**Comment by SamuelEnglard** [comment]
That's what I get for thinking of an idea, never getting around to suggesting it. And then remembering months later and forgetting to check again if it had been suggested. Thanks!

**Comment by hackcraft** [comment]
Never mind, I do check for duplicates and I still end up asking something that's been asked before, or submitting a PR that does the same as another.

**Comment by jnm2** [comment]
I assumed this meant Dictionary.GetOrAdd was added, but I'm not finding it or an IReadOnlyDictionary extension method. If they have been skipped, can this issue be reopened?

**Comment by jnm2** [comment]
Extension method would be on IDictionary rather.

**Comment by svick** [comment]
@jnm2 As far as I can tell, @SamuelEnglard closed this issue in favor of https://github.com/dotnet/corefx/issues/1942, which was about `TryAdd`, but also discussed `GetOrAdd`. But then https://github.com/dotnet/corefx/issues/1942 was closed after `TryAdd` was added.

So yes, I think it makes sense to repoen this issue.

**Comment by jnm2** [comment]
Ugh, I can't believe this missed .NET Standard 2.0. @karelz no exceptions?

**Comment by jnm2** [comment]
Specifically:

```c#
public static TValue GetOrAdd<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, TKey key, TValue value)
{
    if (dictionary == null) throw new ArgumentNullException(nameof(dictionary));

    if (!dictionary.TryGetValue(key, out var r))
        dictionary.Add(key, r = value);
    return r;
}

public static TValue GetOrAdd<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, TKey key, Func<TKey, TValue> valueFactory)
{
    if (dictionary == null) throw new ArgumentNullException(nameof(dictionary));
    if (valueFactory == null) throw new ArgumentNullException(nameof(valueFactory));

    if (!dictionary.TryGetValue(key, out var r))
        dictionary.Add(key, r = valueFactory.Invoke(key));
    return r;
}
```

**Comment by karelz** [comment]
This needs API review. I don't want to rush API reviews. That is just begging for troubles later - you miss things and then you live with your mistakes forever.

It is convenience API, we lived without it for quite some time, we can IMO survive one more release. It is not the end of the world after all. It does not block scenarios.

Probably not the answer you expected, but this is how I feel about it. We should IMO focus primarily on the really impactful things. Makes sense?

**Comment by jnm2** [comment]
It is indeed the answer I expected. 😄

**Comment by jnm2** [comment]
(Don't worry about me, I'm not the type that goes around second-guessing the team's decisions. Fyi.)

**Comment by karelz** [comment]
Thanks for understanding!
IMO it is not that much about decisions, but us being open about prioritization. To ship high-quality SW, we gotta prioritize. And we gotta stabilize.

**Comment by jamesqo** [comment]
@jnm2 I don't think it would be a good idea to add `GetOrAdd` as an extension method on `IDictionary`-- it would give the impression that only 1 lookup is being made. I think it's best just to make this proposal about `Dictionary<K, V>`.

@karelz This is not necessarily a convenience API. For `Dictionary`, this helps avoid duplicate hashtable lookups, and the last time I checked I think I saw around 50 non-test instances of

```cs
if (!dict.ContainsKey(key))
{
    dict.Add(key, ...);
}
```

in corefx.

**Comment by jnm2** [comment]
@jamesqo Yeah, okay. After the BCL has it for a while and we get default interface methods it might make more sense.
